File: frontend/src/interfaces/calculator.ts
export interface Calculator {
    state: CalculatorState;
    isLoading: boolean;
    error: string | null;
    inputDigit: (digit: string) => void;
    inputDecimal: () => void;
    chooseOperation: (operation: Operator) => void;
    toggleSign: () => void;
    backspace: () => void;
    calculate: () => Promise<void>;
    clear: () => void;
}

export interface CalcRequest {
    operand1: number;
    operand2: number;
    operation: Operator;
}

export interface CalcResponse {
    result: number;
}

export interface HistoryEntry {
    id: number;
    operand1: number;
    operand2: number;
    operation: Operator;
    result: number;
}

export interface CalculatorState {
    operand1: number;
    operand2: number;
    operation: Operator;
    result: number | null;
    displayValue: string;
    activeOperand: CalculatorOperand;
    shouldResetDisplay: boolean;
}

export type CalculatorOperand = "operand1" | "operand2";

export type Operator =
    | "add"
    | "subtract"
    | "multiply"
    | "divide"
    | "exponent"
    | "square"
    | "squareRoot"
    | "percent";

File: backend/internal/calculator/service/calculator.go
package service

import (
	"math"

	"backend/sezzle_test/internal/calculator/model"
)

type CalculatorService struct{}

func NewCalculatorService() *CalculatorService {
	return &CalculatorService{}
}

func (s *CalculatorService) Calculate(operation model.Operation, operand1, operand2 float64) (float64, error) {
	switch operation {
	case model.Add:
		return operand1 + operand2, nil
	case model.Subtract:
		return operand1 - operand2, nil
	case model.Multiply:
		return operand1 * operand2, nil
	case model.Divide:
		if operand2 == 0 {
			return 0, ErrDivisionByZero
		}
		return operand1 / operand2, nil
	case model.Exponent:
		return math.Pow(operand1, operand2), nil
	case model.SquareRoot:
		if operand1 < 0 {
			return 0, ErrNegativeRoot
		}
		return math.Sqrt(operand1), nil
	case model.Percent:
		return (operand1 / 100) * operand2, nil
	default:
		return 0, ErrInvalidOperation
	}
}
